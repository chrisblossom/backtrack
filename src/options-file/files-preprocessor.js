/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';

import type { Files } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: Files,
    dirname: string,
};

function filesPreprocessor({ value, dirname }: Args = {}) {
    const resolveSrc = toArray(value).map((files) => {
        if (typeof files === 'boolean') {
            return files;
        }

        if (!files.src) {
            return files;
        }

        const src = path.resolve(dirname, files.src);

        return {
            ...files,
            src,
        };
    });

    return resolveSrc;
}

export { filesPreprocessor };
