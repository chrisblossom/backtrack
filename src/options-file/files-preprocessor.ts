import path from 'path';
import { toArray } from '../utils/object-utils';

import { isCopyFileOptions, Files } from '../types';

type Args = {
	value?: Files;
	dirname?: string;
};

function filesPreprocessor({ value, dirname = '' }: Args = {}) {
	const resolveSrc = toArray(value).map((files) => {
		if (typeof files === 'boolean') {
			return files;
		}

		if (isCopyFileOptions(files)) {
			return files;
		}

		const src = path.resolve(dirname, files.src);

		return {
			...files,
			src,
		};
	});

	return resolveSrc;
}

export { filesPreprocessor };
