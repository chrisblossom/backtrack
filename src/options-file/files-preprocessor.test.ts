import path from 'path';

const filesPreprocessor = (args: any) =>
	require('./files-preprocessor').filesPreprocessor(args);

describe('filesPreprocessor', () => {
	const cwd = process.cwd();

	afterEach(() => {
		process.chdir(cwd);
	});

	test('resolves src file', () => {
		const dir = path.resolve(
			__dirname,
			'../file-manager/__sandbox__/stats1/',
		);
		process.chdir(dir);

		const value = [
			{
				src: 'file1.js',
				dest: 'one.js',
			},
			{
				src: path.resolve(dir, 'file2.js'),
				dest: 'two.js',
			},
			{
				src: 'random.js',
				dest: 'three.js',
				allowChanges: true,
			},
			{
				allowChanges: true,
			},
			{
				makeDirs: ['static'],
			},
		];

		const result = filesPreprocessor({ value, dirname: dir });

		expect(result).toMatchSnapshot();
	});
});
