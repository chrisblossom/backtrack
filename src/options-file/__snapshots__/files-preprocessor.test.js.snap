// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filesPreprocessor resolves src file 1`] = `
Array [
  Object {
    "dest": "one.js",
    "src": "<PROJECT_ROOT>/file1.js",
  },
  Object {
    "dest": "two.js",
    "src": "<PROJECT_ROOT>/file2.js",
  },
  Object {
    "allowChanges": true,
    "dest": "three.js",
    "src": "<PROJECT_ROOT>/random.js",
  },
  Object {
    "allowChanges": true,
  },
  Object {
    "makeDirs": Array [
      "static",
    ],
  },
]
`;
